### PR TITLE
Add configuration.removeOnSession()

### PR DESCRIPTION
**Pull request by robinmacharg**
## Goal

To add the `{add,remove}OnSession()` methods to `Bugsnag` and `BugsnagConfiguration` to bring it in line with other notifiers.

## Design

Simple.  The relevant property is an ObjC array; all necessary checks are performed by the type system and array implementation.  Testing ensures that the method performs as expected.

## Changeset

Two methods added to `Bugsnag` and `BugsnagConfiguration` and tests.

## Tests

Quite comprehensive testing of both addition and removal of blocks.

### Outstanding Questions

- This pull request is ready for:
  - [ ] Initial review of the intended approach, not yet feature complete
  - [ ] Structural review of the classes, functions, and properties modified
  - [X] Final review

- [ ] The correct target branch has been selected (`master` for fixes, `next` for
  features)
- [ ] Consistency across platforms for structures or concepts added or modified
- [X] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [X] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [X] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: Source/Bugsnag.m [hunk 1]
```diff
@@ -318,6 +318,11 @@ + (void)setUser:(NSString *_Nullable)userId
     [[self configuration] setUser:userId withName:name andEmail:email];
 }
 
++ (void)removeOnSessionBlock:(BugsnagOnSessionBlock _Nonnull )block
```
Comment (tomlongridge) [inline Source/Bugsnag.m]: We need an equivalent `addOnSessionBlock` in Client too.

File: Source/Bugsnag.h [hunk 2]
```diff
@@ -327,4 +327,12 @@ static NSString *_Nonnull const BugsnagSeverityInfo = @"info";
        withName:(NSString *_Nullable)name
        andEmail:(NSString *_Nullable)email;
```
[the remark this answers is not in the document]
Comment (robinmacharg) [reply Source/Bugsnag.h]: Missed from the underlying task.  Now added.

File: Source/Bugsnag.m (hunk 1 above)
Comment (robinmacharg) [reply Source/Bugsnag.m]: Missed from the underlying task.  Now added.

File: Source/Bugsnag.m [hunk 3]
```diff
@@ -318,6 +318,11 @@ + (void)setUser:(NSString *_Nullable)userId
...
+{
+    [[self configuration] removeOnSessionBlock:block];
```
[the remark this answers is not in the document]
Comment (robinmacharg) [reply Source/Bugsnag.m]: This has been tracked internally, with a link back to this comment.

File: Source/BugsnagConfiguration.m [hunk 4]
```diff
@@ -193,6 +193,10 @@ - (void)addOnSessionBlock:(BugsnagOnSessionBlock)block {
     [(NSMutableArray *)self.onSessionBlocks addObject:[block copy]];
 }
 
+- (void)removeOnSessionBlock:(BugsnagOnSessionBlock)block {
+    [(NSMutableArray *)self.onSessionBlocks removeObject:block];
```
[the remark this answers is not in the document]
Comment (robinmacharg) [reply Source/BugsnagConfiguration.m]: Yes, if we push the access/iteration down in to Configuration.  It's used [here](https://github.com/bugsnag/bugsnag-cocoa/blob/3d35ed78052d4048f4f3ce1445ca4f4ca258b793/Source/BugsnagSessionTrackingApiClient.m#L46).  I'd prefer to do that as a separate task.

File: Tests/BugsnagConfigurationTests.m [hunk 5]
```diff
@@ -1,33 +1,245 @@
...
+    // Check it's called on new session start
+    [Bugsnag pauseSession];
+    called++;
+    [Bugsnag startSession];
+    [self waitForExpectations:@[expectation2] timeout:1.0];
+
+    // Check it's NOT called once the block's deleted
+    [Bugsnag pauseSession];
+    called++;
+    [config removeOnSessionBlock:sessionBlock];
```
[the remark this answers is not in the document]
Comment (robinmacharg) [reply Tests/BugsnagConfigurationTests.m]: "Should have", but at present does.  I'll leave as-is with a note added to the task to move config usage to a deep-copying version.
